Added a new blog post about go and opentelemetry (#1581)
open-telemetry/opentelemetry.io: content/en/blog/2022/how-to-get-started-with-opentelemetry-go.md

It is a CNCF community-driven open-source project (Cloud Native Computing Foundation, the folks in charge of Kubernetes).

In a cloud-native environment, we use OpenTelemetry (OTel for short) to gather data from our system operations and events. In other words, to instrument our distributed services. This data enables us to understand and investigate our software’s behavior and troubleshoot performance issues and errors.

OpenTelemetry serves as a standard observability framework that captures all data under a single specification. It provides several components, including:

1. APIs and SDKs per programming language for generating telemetry
2. The OpenTelemetry Collector; receives, processes, and exports telemetry data to different destinations.
3. [OTLP](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md) protocol for shipping telemetry data
   To get a deeper understanding of this technology, including its structure and the primary motivation, visit this guide.

For this OpenTelemetry Golang guide, here are the terms you need to know:

> - **Span:** The most basic unit. A span represents an event in our system (e.g., an HTTP request or a database operation that spans over time). A span would usually be the parent of another span, its child, or both.

Reviewer: I'd recommend sticking with the definitions we have in our glossary on the website here. It's fine to copy/paste those definitions, but what you have here does differ a bit from the glossary, which is considered canonical.

FWIW if you see something in the glossary and thing, "hmmm, this could improve" ... please send a PR! I'm more than happy to improve the glossary and incorporate some of what you have here if you think it will be helpful.

Author: Hi, Sorry for the late reply. I will check all of your feedback ASAP

- **Exporter:** Once we create a span, the exporter handles sending the data to our backend (e.g., in memory, Jaeger Tracing, or console output)
- **Context propagation:** The mechanism that allows us to correlate events across distributed services. Context is referred to as the metadata we collect and transfer. Propagation is how the context is packaged and transferred across services, often via HTTP headers. Context propagation is one of the areas where OpenTelemetry shines.
- **Instrumentation:** instrumentation libraries gather the data and generate spans based on different libraries in our applications (Kafka, Mongo, Gin, etc.).

There are two ways to instrument our app – manually and automatically:

- **Auto instrumentation:** Automatically create spans from the application libraries we use with ready-to-use OpenTelemetry libraries.
- **Manual instrumentation:** Manually add code to your application to define the beginning and end of each span and the payload.

>   To understand more of the OpenTelemetry jargon, [visit the official documentation](https://opentelemetry.io/docs/concepts/data-sources/).

Reviewer 2: ```suggestion

  To learn more, see [Instrumenting](/docs/concepts/instrumenting/).
```

The original link is breaking the build.

Reviewer 2: I'm going to commit this suggestion so that we can get the build to complete.

## Hello world: OpenTelemetry Go example

We will start by creating our to-do service and installing two libraries (Gin and Mongo) to understand how instrumentations work.

### Step 1: Create main.go file for our to-do app

1. Install Gin and Mongo-driver

```go
go get -u github.com/gin-gonic/gin
go get go.mongodb.org/mongo-driver/mongo
```

2. Set up gin and mongo to listen on “/todo”

3. Create some to-do’s to seed Mongo

```go
package main
import (

> "context"

Reviewer: formatting got borked here, this code sample needs reformating"

}
func setupWebServer() {
r := gin.Default()
r.GET("/todo", func(c *gin.Context) {
collection := client.Database("todo").Collection("todos")
//Important: Make sure to pass c.Request.Context() as the context and not c itself - TBD
cur, findErr := collection.Find(c.Request.Context(), bson.D{})
if findErr != nil {
c.AbortWithError(500, findErr)
return
}
results := make([]interface{}, 0)
curErr := cur.All(c, &results)
if curErr != nil {
c.AbortWithError(500, curErr)
return
}
c.JSON(http.StatusOK, results)
})
\_ = r.Run(":8080")
}
```

Now that our small todo app is ready, let’s introduce OpenTelemetry.

### Step 2: Install OpenTelemetry Go

We will be configuring OpenTelemetry to instrument our Go app.

1. To install the OTel SDK, run:

```go
go get go.opentelemetry.io/otel /
go.opentelemetry.io/otel/sdk /
```

2. Instrument our Gin and Mongo libraries to generate traces.

3. Gin instrumentation: Install otelgin

```go
go get go.opentelemetry.io/contrib/instrumentation/github.com/gin-gonic/gin/otelgin
```

4. Mongo instrumentation: Install otelmongo

```go
go get go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo

> ```

Reviewer: I'd recommend collapsing this into a single call to `go get` than 4 separate steps for people to write.

Also the language chosen for the code blocks should be `console` and not `go`.

### Gin instrumentation: gin.Context

We previously discussed the idea of context propagation – the way to transfer metadata between distributed services to correlate events in our system.

The Gin framework has its own type gin.Context which gets passed as a parameter to an HTTP handler. However, the context that should be passed down to the mongo operations is the standard Go library Context object, available in gin.Context.Request.Context.

```go
//Make sure to pass c.Request.Context() as the context and not c itself
cur, findErr := collection.Find(c.Request.Context(), bson.D{})

> So make sure that you pass the Context to the mongodb operation. Check out this issue for more info.

Reviewer: This should be outside the code block.

```

We now have our todo app ready and instrumented. It’s time to utilize OpenTelemetry to its full potential. Our ability to visualize traces is where the true troubleshooting power of this technology comes into play.

For visualization, we’ll be using the open-source Jaeger Tracing and Aspecto.

## Visualization with Jaeger

The setup to export traces to Jaeger or Aspecto is relatively similar. Follow along with the Jaeger setup, then switch to Aspecto by changing a few lines of code.

> ### OpenTelemetry Go and Jaeger Tracing: Export traces to Jaeger

Reviewer: Now that Jaeger supports OTLP natively, should this section change to just set up the OTLP exporter and then send to jaeger that way?

Author: Thats a great idea, we will update the blog post in the future =)

[Jaeger Tracing](https://www.aspecto.io/blog/jaeger-tracing-the-ultimate-guide/) is a suite of open source projects managing the entire distributed tracing “stack”: client, collector, and UI. Jaeger UI is the most commonly used open-source to visualize traces.

Here’s what the setup looks like:

1. Install the Jaeger exporter

```go
go get go.opentelemetry.io/otel/exporters/jaeger
```

2. Create a tracing folder and a jaeger.go file

3. Add the following code to the file

```go
package tracing
import (
"go.opentelemetry.io/otel/exporters/jaeger"
"go.opentelemetry.io/otel/sdk/resource"
sdktrace "go.opentelemetry.io/otel/sdk/trace"
semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
)

> func JaegerTraceProvider() (\*sdktrace.TracerProvider, error) {

Reviewer: ```suggestion
func JaegerTraceProvider() (*sdktrace.TracerProvider, error) {
```

connectMongo()
setupWebServer()
}
```

Next, we are going to hook up the instrumentations we installed.

5. Add the Mongo instrumentation. In our connectMongo function by adding this line

```go
opts.Monitor = otelmongo.NewMonitor()
```

The function shold look like this

```go
func connectMongo() {
opts := options.Client()
//Mongo OpenTelemetry instrumentation
opts.Monitor = otelmongo.NewMonitor()
opts.ApplyURI("mongodb://localhost:27017")
client, \_ = mongo.Connect(context.Background(), opts)
//Seed the database with some todo's
docs := []interface{}{
bson.D{{"id", "1"}, {"title", "Buy groceries"}},
bson.D{{"id", "2"}, {"title", "install Aspecto.io"}},
bson.D{{"id", "3"}, {"title", "Buy dogz.io domain"}},
}
client.Database("todo").Collection("todos").InsertMany(context.Background(), docs)
}
```

Now, add the Gin instrumentation.

6. Go to the startWebServer function and add this line right after we create the gin instance

```go
r.Use(otelgin.Middleware("todo-service"))
```

The function should look like this

```go
func startWebServer() {
r := gin.Default()
//Gin OpenTelemetry instrumentation
r.Use(otelgin.Middleware("todo-service"))

> r.GET("/todo", func(c \*gin.Context) {

Reviewer: ```suggestion
r.GET("/todo", func(c *gin.Context) {
```
This and all other `*` need to be un-escaped.

---

> title: Go Web App Instrumentation

Reviewer 2: Rename the file to more closely match the title, so something like: `go-web-app-instrumentation.md`.

linkTitle: Go App Instrumentation
date: 2022-06-27
author: Naveh Mevorach (Aspecto)
canonical_url: https://www.aspecto.io/blog/opentelemetry-go-getting-started/
---

In this blog post, you will learn hands-on how to create and visualize traces with OpenTelemetry Go without prior knowledge.

We will start with creating a simple to-do app that uses Mongo and the Gin framework. Then, we will send tracing data to Jaeger Tracing for visualization. You can find all the relevant files in this [Github repository](https://github.com/aspecto-io/opentelemetry-examples/tree/master/go).

> !["OpenTelemetry Go The Mandalorian"](https://www.aspecto.io/wp-content/uploads/2022/06/OpenTelemetry-Go-The-Mandalorian-2048x1406.png)

Reviewer 3: @chalin, @austinlparker : should images in the blog/docs live here or is it okay to load them from a 3rd party URL?

Reviewer 2: Ideally they should live in this repo, and (double) ideally they should be packaged as [page resources]. 

@NavehMevorach, if you add the images to this PR, I can organize them all into a [page bundle] for this post for you.

[page resources]: https://gohugo.io/content-management/page-resources/
[page bundle]: https://gohugo.io/content-management/page-bundles/

Author: @chalin Let me first regenerate the images and then I will add them to the PR

Reviewer 2: @NavehMevorach - if you'd like I can repackage things into a page bundle after you're done and before the final publication.

## What to Expect

- Intro to OpenTelemetry
- Hello world: OpenTelemetry GO example
  - Create main.go file with Gin and Mongo
  - Install OpenTelemetry GO
  - Gin instrumentation: gin.Context

> - Visualization with Jaeger

Reviewer 2: This "What to expect" section seems to essentially be a Table of Contents. I'd drop it given that a real ToC with hyperlinks is autogenerated. WDYT?

- **Context propagation:** Allows all Data Sources to share an underlying context mechanism for storing state and accessing data across the lifespan of a Transaction. See context propagation spec.
- **Instrumentation:** Denotes the Library that provides the instrumentation for a given Instrumented Library. Instrumented Library and Instrumentation Library may be the same Library if it has built-in OpenTelemetry instrumentation.

There are two ways to instrument our app – manually and automatically:

- **Auto instrumentation:** Automatically create spans from the application libraries we use with ready-to-use OpenTelemetry libraries.
- **Manual instrumentation:** Manually add code to your application to define the beginning and end of each span and the payload.

  To learn more, see [Instrumenting](/docs/concepts/instrumenting/).

## Hello world: OpenTelemetry Go example

We will start by creating our to-do service and installing two libraries (Gin and Mongo) to understand how instrumentations work.

### Step 1: Create main.go file for our to-do app

1. Install Gin and Mongo-driver

```shell
go get -u github.com/gin-gonic/gin
go get go.mongodb.org/mongo-driver/mongo

> ```

Reviewer 2: All of the code blocks that are part of lists need to be indented. Fix this here and below.

2. Set up gin and mongo to listen on “/todo”

3. Create some to-do’s to seed Mongo

```go
package main
import (
    "context"
    "net/http"
    "github.com/gin-gonic/gin"
    "go.mongodb.org/mongo-driver/bson"
    "go.mongodb.org/mongo-driver/mongo"
    "go.mongodb.org/mongo-driver/mongo/options"
)
var client * mongo.Client
func main() {
    connectMongo()
    setupWebServer()

> }

Reviewer 2: Consider running a formatter over all the code code excerpts.
At a minimum, add a blank line between elements.

    }
    client.Database("todo").Collection("todos").InsertMany(context.Background(), docs)
}
func startWebServer() {
    r: = gin.Default()
    //gin OpenTelemetry instrumentation
    r.Use(otelgin.Middleware("todo-service"))
    r.GET("/todo", func(c * gin.Context) {
        collection: = client.Database("todo").Collection("todos")
        //Make sure to pass c.Request.Context() as the context and not c itself
        cur, findErr: = collection.Find(c.Request.Context(), bson.D {})
        if findErr != nil {
            c.AbortWithError(500, findErr)
            return
        }
        results: = make([] interface {}, 0)
        curErr: = cur.All(c, & results)
        if curErr != nil {
            c.AbortWithError(500, curErr)
            return
        }
        c.JSON(http.StatusOK, results)
    })
    _ = r.Run(":8080")
}
```

### Export traces to Jaeger

1. Run the todo-service with go run main.go
2. Make an HTTP GET request to localhost:8080/todo to generate some traces in Go
3. Open Jaeger at http://localhost:16686/search to view those traces

You can now see the Jaeger UI. Select todo-service and click on Find traces. You should see your trace on the right:

> !["Jaeger UI displays opentelemetry traces in go for our todo-service"](https://lh5.googleusercontent.com/ZeFbAE9-XVSc-5GHjZkslHuJ3f01VQqSrObOgLY9yDSjuTyJdvvAzIapIvTQqumFTUP2BZE4gxd-Vt2JXjvqO1ep3JUBhkHKiry_m8bSAwwvEf3kKNfzFiKwzFP8E3btWtQV0pLZZWnsbY-sUA)

Reviewer 2: This and the next image are difficult to read. Could you regenerate some images but make the window that you're taking a screenshot of narrower?

---

> title: Go Web App & PpenTelemetry Instrumentation

Reviewer 2: There's a typo in the title: PpenTelemetry.

How about the shorter title of:

# Go Web App Instrumentation

Reviewer 2: Oh, I see that you've put that as the `linkTitle`. If you agree with the change mentioned above, then drop the `linkTitle` field.

Reviewer 2: Or maybe the `linkTitle` could be just: **Go App Instrumentation**

linkTitle: Go Web App Instrumentation
date: 2022-06-27
author: Naveh Mevorach (Aspecto)
canonical_url: https://www.aspecto.io/blog/opentelemetry-go-getting-started/
---

> # How to Get Started with OpenTelemetry Go

Reviewer 2: Remove this level 1 heading

> In this guide, you will learn hands-on how to create and visualize traces with OpenTelemetry Go without prior knowledge.

Reviewer 2: ```suggestion
In this blog post, you will learn hands-on how to create and visualize traces with OpenTelemetry Go without prior knowledge.
```

- **Context propagation:** Allows all Data Sources to share an underlying context mechanism for storing state and accessing data across the lifespan of a Transaction. See context propagation spec.
- **Instrumentation:** Denotes the Library that provides the instrumentation for a given Instrumented Library. Instrumented Library and Instrumentation Library may be the same Library if it has built-in OpenTelemetry instrumentation.

There are two ways to instrument our app – manually and automatically:

- **Auto instrumentation:** Automatically create spans from the application libraries we use with ready-to-use OpenTelemetry libraries.
- **Manual instrumentation:** Manually add code to your application to define the beginning and end of each span and the payload.

  To learn more, see [Instrumenting](/docs/concepts/instrumenting/).

## Hello world: OpenTelemetry Go example

We will start by creating our to-do service and installing two libraries (Gin and Mongo) to understand how instrumentations work.

### Step 1: Create main.go file for our to-do app

1. Install Gin and Mongo-driver

> ```go

Reviewer 2: ```suggestion
```shell
```

                }
            },
            bson.D {
                {
                    "id", "3"
                }, {
                    "title", "Buy dogz.io domain"
                }
            },
    }
    client.Database("todo").Collection("todos").InsertMany(context.Background(), docs)
}
func setupWebServer() {
    r: = gin.Default()
    r.GET("/todo", func(c * gin.Context) {
        collection: = client.Database("todo").Collection("todos")
        //Important: Make sure to pass c.Request.Context() as the context and not c itself - TBD
        cur, findErr: = collection.Find(c.Request.Context(), bson.D {})
        if findErr != nil {
            c.AbortWithError(500, findErr)
            return
        }
        results: = make([] interface {}, 0)
        curErr: = cur.All(c, & results)
        if curErr != nil {
            c.AbortWithError(500, curErr)
            return
        }
        c.JSON(http.StatusOK, results)
    })
    _ = r.Run(":8080")
}
```

Now that our small todo app is ready, let’s introduce OpenTelemetry.

### Step 2: Install OpenTelemetry Go

We will be configuring OpenTelemetry to instrument our Go app.

1. To install the OTel SDK, run:

> ```console

Reviewer 2: Only use `console` when you include command-line prompts and/or wish command output to be highlighted differently.

```suggestion
```shell
```

go get go.opentelemetry.io/otel /
go.opentelemetry.io/otel/sdk /
```

2. Instrument our Gin and Mongo libraries to generate traces.

3. Gin & Mongo instrumentation: Install otelgin & otelmongo

> ```console

Reviewer 2: ```suggestion
```shell
```

```go
//Make sure to pass c.Request.Context() as the context and not c itself
cur, findErr := collection.Find(c.Request.Context(), bson.D{})
```
So make sure that you pass the Context to the mongodb operation. Check out this issue for more info.

We now have our todo app ready and instrumented. It’s time to utilize OpenTelemetry to its full potential. Our ability to visualize traces is where the true troubleshooting power of this technology comes into play.

For visualization, we’ll be using the open-source Jaeger Tracing.

## Visualization with Jaeger

### OpenTelemetry Go and Jaeger Tracing: Export traces to Jaeger

[Jaeger Tracing](https://www.aspecto.io/blog/jaeger-tracing-the-ultimate-guide/) is a suite of open source projects managing the entire distributed tracing “stack”: client, collector, and UI. Jaeger UI is the most commonly used open-source to visualize traces.

Here’s what the setup looks like:

1. Install the Jaeger exporter

> ```console

Reviewer 2: ```suggestion
```shell
```

---
title: Instrument your Go web application with openTelemetry

> linkTitle: Get Started with OpenTelemetry & Go

Reviewer 3: This needs to be changed as well and also short & crisp, i.e. Go Web App Instrumentation?

date: 2022-06-27

> author: Team Aspecto

Reviewer 3: sorry for being so pedantic, but this is another one I'd like to call out: Until now we have blog posts from individuals or community teams, but not a vendor. I would prefer having the name of the author here, with a link to their github profile, or we can do `author: Naveh Mevorach (Aspecto)` ? (cc: @cartermp, @chalin, @austinlparker)

Reviewer 3: @NavehMevorach, please give the others some time to respond as well.

Reviewer 2: +1

canonical_url: https://www.aspecto.io/blog/opentelemetry-go-getting-started/
---

# How to Get Started with OpenTelemetry Go

In this guide, you will learn hands-on how to create and visualize traces with OpenTelemetry Go without prior knowledge.

> We will start with creating a simple to-do app that uses Mongo and the Gin framework. Then, we will send tracing data to Jaeger Tracing and to Aspecto for visualization. You can find all the relevant files in this [Github repository](https://github.com/aspecto-io/opentelemetry-examples/tree/master/go).

Reviewer 3: remove "and to Aspecto"

> !["OpenTelemetry Go The Mandalorian"](https://www.aspecto.io/wp-content/uploads/2022/06/OpenTelemetry-Go-The-Mandalorian-2048x1406.png)

Reviewer 3: I have my issues with this picture, since The Mandalorian is a trademark, and there's probably also some copyrights on the images.

Author: @svrnm , Hi! Do you mind to clarify your concerns regarding that illustration? Our designer made it so there aren't any copyright issues in that aspect.

Reviewer 3: I have no legal expertise to make a clear decision if this picture (and its name & description) are OK the way they are presented here or who is liable if they are not (Linux Foundation, Aspecto, your designer?), so I rather prefer not having it. @cartermp, @chalin, @austinlparker: thoughts?

Reviewer 2: I will ask here at the CNCF and get back to you.

Reviewer 2: @caniszczyk says that it's ok if @NavehMevorach wants to use this image.

Reviewer 3: thanks for clarification, @chalin & @caniszczyk!

Reviewer 2: Thanks for bring up the issue.

## What to Expect

- Intro to OpenTelemetry
- Hello world: OpenTelemetry GO example
  - Create main.go file with Gin and Mongo
  - Install OpenTelemetry GO
  - Gin instrumentation: gin.Context
- Visualization with Jaeger
  - OpenTelemetry Go and Jaeger Tracing

> ## Intro to OpenTelemetry

Reviewer 3: This section is not required, please point to the existing concept docs that give you that description.

Reviewer 3: @NavehMevorach you unresolved this conversation, but I don't see it resolved. what has changed?

Author: I'm not sure what you want me to remove. The title itself?

Reviewer 3: the whole section, everything from line 24 to 53: this is a blog on opentelemetry.io, so either people know about OpenTelemetry already or they should read the [Concepts](https://opentelemetry.io/docs/concepts/) to learn more

go get go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo
```
### Gin instrumentation: gin.Context

We previously discussed the idea of context propagation – the way to transfer metadata between distributed services to correlate events in our system.

The Gin framework has its own type gin.Context which gets passed as a parameter to an HTTP handler. However, the context that should be passed down to the mongo operations is the standard Go library Context object, available in gin.Context.Request.Context.

```go
//Make sure to pass c.Request.Context() as the context and not c itself
cur, findErr := collection.Find(c.Request.Context(), bson.D{})
```
So make sure that you pass the Context to the mongodb operation. Check out this issue for more info.

We now have our todo app ready and instrumented. It’s time to utilize OpenTelemetry to its full potential. Our ability to visualize traces is where the true troubleshooting power of this technology comes into play.

For visualization, we’ll be using the open-source Jaeger Tracing and Aspecto.

## Visualization with Jaeger

> The setup to export traces to Jaeger or Aspecto is relatively similar. Follow along with the Jaeger setup, then switch to Aspecto by changing a few lines of code.

Reviewer 3: remove "Aspecto"

        if curErr != nil {
            c.AbortWithError(500, curErr)
            return
        }
        c.JSON(http.StatusOK, results)
    })
    _ = r.Run(":8080")
}
```

### Export traces to Jaeger

1. Run the todo-service with go run main.go
2. Make an HTTP GET request to localhost:8080/todo to generate some traces in Go
3. Open Jaeger at http://localhost:16686/search to view those traces

You can now see the Jaeger UI. Select todo-service and click on Find traces. You should see your trace on the right:
!["Jaeger UI displays opentelemetry traces in go for our todo-service"](https://lh5.googleusercontent.com/ZeFbAE9-XVSc-5GHjZkslHuJ3f01VQqSrObOgLY9yDSjuTyJdvvAzIapIvTQqumFTUP2BZE4gxd-Vt2JXjvqO1ep3JUBhkHKiry_m8bSAwwvEf3kKNfzFiKwzFP8E3btWtQV0pLZZWnsbY-sUA)

Jaeger UI displays opentelemetry traces in go for our todo-service
By clicking the trace, you can drill down and see more details about it that allow you to further investigate on your own:
!["Jaeger UI. To-do service drill down."](https://lh5.googleusercontent.com/5KI-tGGriWaMf98vNjewZZTwE1f-g7dQJXCEaCWmklT_xmCc5E_2VSGcRDeKf4GNZwRSNnSpQCQFH-1nUXIF7a5gd6Y7odFiEukSbaWaukFKP0cXXylHIqGJvAMfbQ2p60nt3wmeOwTtRr3eKQ)

## Summary

> That’s all folks! We hope this guide was informative and easy to follow. You can find all files ready to use in our Github repository.

Reviewer 3: link to the github repo.

> If you have any questions or issues, feel free to reach out to us via our chat.

Reviewer 3: Where is our chat? Link to CNCF slack #opentelemetry?

> To learn more about it, check out our OpenTelemetry Bootcamp. It’s a 6 episode YouTube series that covers OpenTelemetry from zero to 100 including the implementation of opentelemetry in production, security, sampling, and more. Completely free and vendor-neutral.

Reviewer 3: Remove everything from line 423 to 431 pointing to vendor content.
